sycl/source/detail/graph_impl.cpp

```diff
@@ -0,0 +1,425 @@
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+
+namespace ext {
+namespace oneapi {
+namespace experimental {
+namespace detail {
+
+namespace {
+
+/// Recursively check if a given node is an exit node, and add the new nodes as
+/// successors if so.
+/// @param[in] CurrentNode Node to check as exit node.
+/// @param[in] NewInputs Noes to add as successors.
+void connect_to_exit_nodes(
+    std::shared_ptr<node_impl> CurrentNode,
+    const std::vector<std::shared_ptr<node_impl>> &NewInputs) {
+  if (CurrentNode->MSuccessors.size() > 0) {
+    for (auto Successor : CurrentNode->MSuccessors) {
+      connect_to_exit_nodes(Successor, NewInputs);
+    }
+
+  } else {
+    for (auto Input : NewInputs) {
+      CurrentNode->register_successor(Input, CurrentNode);
+    }
+  }
+}
+
+/// Recursive check if a graph node or its successors contains a given kernel
+/// argument.
+/// @param[in] Arg The kernel argument to check for.
+/// @param[in] CurrentNode The current graph node being checked.
+/// @param[in,out] Deps The unique list of dependencies which have been
+/// identified for this arg.
+/// @return True if a dependency was added in this node or any of its
+/// successors.
+bool check_for_arg(const sycl::detail::ArgDesc &Arg,
+                   const std::shared_ptr<node_impl> &CurrentNode,
+                   std::set<std::shared_ptr<node_impl>> &Deps) {
+  bool SuccessorAddedDep = false;
+  for (auto &Successor : CurrentNode->MSuccessors) {
+    SuccessorAddedDep |= check_for_arg(Arg, Successor, Deps);
+  }
+
+  if (!CurrentNode->is_empty() && Deps.find(CurrentNode) == Deps.end() &&
+      CurrentNode->has_arg(Arg) && !SuccessorAddedDep) {
+    Deps.insert(CurrentNode);
+    return true;
+  }
+  return SuccessorAddedDep;
+}
+} // anonymous namespace
+
+void exec_graph_impl::schedule() {
+  if (MSchedule.empty()) {
+    for (auto Node : MGraphImpl->MRoots) {
+      Node->topology_sort(Node, MSchedule);
+    }
+  }
+}
+
+std::shared_ptr<node_impl> graph_impl::add_subgraph_nodes(
+    const std::list<std::shared_ptr<node_impl>> &NodeList) {
+  // Find all input and output nodes from the node list
+  std::vector<std::shared_ptr<node_impl>> Inputs;
+  std::vector<std::shared_ptr<node_impl>> Outputs;
+  for (auto &NodeImpl : NodeList) {
+    if (NodeImpl->MPredecessors.size() == 0) {
+      Inputs.push_back(NodeImpl);
+    }
+    if (NodeImpl->MSuccessors.size() == 0) {
+      Outputs.push_back(NodeImpl);
+    }
+  }
+
+  // Recursively walk the graph to find exit nodes and connect up the inputs
+  // TODO: Consider caching exit nodes so we don't have to do this
+  for (auto NodeImpl : MRoots) {
+    connect_to_exit_nodes(NodeImpl, Inputs);
+  }
+
+  return this->add(Outputs);
+}
+
+sycl::event
+exec_graph_impl::exec(const std::shared_ptr<sycl::detail::queue_impl> &Queue) {
+  sycl::event RetEvent = enqueue(Queue);
+  // TODO: Remove this queue wait. Currently waiting on the event returned from
+  // graph execution does not work.
+  Queue->wait();
```

Comment:
Would prefer not to merge this blocking wait into mainline as we fixed it in https://github.com/reble/llvm/pull/190

Reply:
done. I merge part of this PR to remove blocking wait. Please have another look.

Reply:
Looks good (I can't resolve comment threads on this PR, so feel free to hit resolve for me)